Fix docker build/run on Linux

**Describe the bug**
Fix Docker stuff on Linux. See https://github.com/fmorency/many-framework/tree/fmorency/fix-docker-compose

**To Reproduce**
Steps to reproduce the behavior:
1. `cd many-framework/docker/e2e`
2. `eval $(ssh-agent)`
3. `DOCKER_BUILDKIT=1 make start-docker-compose`

```
panic: could not create directory "/tendermint/config": mkdir /tendermint/config: permission denied
```

**Expected behavior**
Docker composition should build and start properly

**Desktop (please complete the following information):**
 - OS: Linux

**Additional context**
This is a permission problem related to the way Docker works in Linux vs how Docker works in MacOS.

--- a/docker/abci/Dockerfile
+++ b/docker/abci/Dockerfile
@@ -3,7 +3,9 @@ FROM many_local/repo_build:latest as builder
 # ==== STAGE 2 ====
 FROM debian:stable as runtime
 
-RUN useradd -ms /bin/bash many
+ARG USER_ID
+
+RUN useradd -u ${USER_ID} -ms /bin/bash many
 
 STOPSIGNAL SIGTERM
 

--- a/docker/e2e/Makefile
+++ b/docker/e2e/Makefile
@@ -2,7 +2,7 @@
 .PHONY: clean
 clean:
 	if [ -f "${PWD}/genfiles/docker-compose.json" ]; then \
-		docker-compose -f genfiles/docker-compose.json down; \
+		docker-compose -p e2e down; \
 	fi
 	rm -rf genfiles/
 
@@ -18,10 +18,10 @@ build-repo:
 	docker build --ssh default --tag many_local/repo_build ../.. --target builder -f ../build.dockerfile
 
 build-ledger: build-repo
-	docker build --ssh default --tag many/ledger ../.. --target runtime -f ../ledger/Dockerfile
+	docker build --build-arg USER_ID=$$(id -u) --ssh default --tag many/ledger ../.. --target runtime -f ../ledger/Dockerfile
 
 build-abci: build-repo
-	docker build --ssh default --tag many/abci ../.. --target runtime -f ../abci/Dockerfile
+	docker build --build-arg USER_ID=$$(id -u) --ssh default --tag many/abci ../.. --target runtime -f ../abci/Dockerfile
 
 .PHONY: start-ledger-local
 start-ledger-local: build-ledger
@@ -42,19 +42,20 @@ NB_NODES ?= 4
 
 genfiles/docker-compose.json: jsonnet-docker
 	docker run -it --rm -v "${PWD}:/volume:ro" bitnami/jsonnet \
- 		/volume/docker-compose.jsonnet --tla-code nb_nodes=$(NB_NODES) > $@
+		/volume/docker-compose.jsonnet --tla-code nb_nodes=$(NB_NODES) > $@
 
 genfiles/node%: tendermint-docker openssl-docker
 	mkdir -p "$@/tendermint"
-	docker run -it --rm -v "${PWD}/$@/tendermint:/tendermint" tendermint/tendermint:v0.35.1 init validator
-	docker run -it --rm -v ${PWD}/$@/:/export alpine/openssl genpkey -algorithm Ed25519 -out /export/ledger.pem
-	docker run -it --rm -v ${PWD}/$@/:/export alpine/openssl genpkey -algorithm Ed25519 -out /export/abci.pem
+	docker run --user $$(id -u) -it --rm -v "${PWD}/$@/tendermint:/tendermint" tendermint/tendermint:v0.35.1 init validator
+	docker run --user $$(id -u) -it --rm -v ${PWD}/$@/:/export alpine/openssl genpkey -algorithm Ed25519 -out /export/ledger.pem
+	docker run --user $$(id -u) -it --rm -v ${PWD}/$@/:/export alpine/openssl genpkey -algorithm Ed25519 -out /export/abci.pem
+	mkdir -p "$@/persistent-ledger"
 	cp ledger_state.json $@/ledger_state.json
 
 
-generate-tendermint-e2e-config:
+generate-e2e-config:
 	I=0; while [[ $$I -lt ${NB_NODES} ]]; do \
-  		echo $$I genfiles/node$$I; \
+		echo $$I genfiles/node$$I; \
 		make genfiles/node$$I; \
 		(( I = I + 1 )); \
 	done

--- a/docker/e2e/Makefile
+++ b/docker/e2e/Makefile
@@ -42,7 +42,7 @@ NB_NODES ?= 4
 
 genfiles/docker-compose.json: jsonnet-docker
 	docker run -it --rm -v "${PWD}:/volume:ro" bitnami/jsonnet \
-		/volume/docker-compose.jsonnet --tla-code nb_nodes=$(NB_NODES) > $@
+		/volume/docker-compose.jsonnet --tla-code nb_nodes=$(NB_NODES) --tla-code user=$$(id -u) > $@
 
 genfiles/node%: tendermint-docker openssl-docker
 	mkdir -p "$@/tendermint"
@@ -68,7 +68,7 @@ start-all-nodes:
 stop-node:
 	docker-compose -f genfiles/docker-compose.json -p e2e down 
 
-start-docker-compose: build-abci build-ledger generate-tendermint-e2e-config genfiles/docker-compose.json start-all-nodes 
+start-docker-compose: build-abci build-ledger generate-e2e-config genfiles/docker-compose.json start-all-nodes
 
 chaos: 
 	sh chaos.sh 

--- a/docker/e2e/docker-compose.jsonnet
+++ b/docker/e2e/docker-compose.jsonnet
@@ -1,9 +1,10 @@
 //
 
-local abci(i) = {
+local abci(i, user) = {
     image: "many/abci",
     ports: [ (8000 + i) + ":8000" ],
     volumes: [ "./node" + i + ":/genfiles:ro" ],
+    user: "" + user,
     command: [
         "--many", "0.0.0.0:8000",
         "--many-app", "http://ledger-" + i + ":8000",
@@ -14,8 +15,9 @@ local abci(i) = {
     depends_on: [ "ledger-" + i ],
 };
 
-local ledger(i) = {
+local ledger(i, user) = {
     image: "many/ledger",
+    user: "" + user,
     volumes: [
         "./node" + i + "/persistent-ledger:/persistent",
         "./node" + i + ":/genfiles:ro",
@@ -29,7 +31,7 @@ local ledger(i) = {
     ],
 };
 
-local tendermint(i, tendermint_tag="v0.35.1") = {
+local tendermint(i, user, tendermint_tag="v0.35.1") = {
     image: "tendermint/tendermint:" + tendermint_tag,
     command: [
         "--log-level", "info",
@@ -37,19 +39,20 @@ local tendermint(i, tendermint_tag="v0.35.1") = {
         "--rpc.laddr", "tcp://0.0.0.0:26657",
         "--proxy-app", "tcp://abci-" + i + ":26658",
     ],
+    user: "" + user,
     volumes: [
         "./node" + i + "/tendermint/:/tendermint"
     ],
     ports: [ "" + (26600 + i) + ":26600" ],
 };
 
-function(nb_nodes=4) {
+function(nb_nodes=4, user=1000) {
     version: '3',
     services: {
-        ["abci-" + i]: abci(i) for i in std.range(0, nb_nodes - 1)
+        ["abci-" + i]: abci(i, user) for i in std.range(0, nb_nodes - 1)
     } + {
-        ["ledger-" + i]: ledger(i) for i in std.range(0, nb_nodes - 1)
+        ["ledger-" + i]: ledger(i, user) for i in std.range(0, nb_nodes - 1)
     } + {
-        ["tendermint-" + i]: tendermint(i) for i in std.range(0, nb_nodes - 1)
+        ["tendermint-" + i]: tendermint(i, user) for i in std.range(0, nb_nodes - 1)
     }
 }

--- a/docker/e2e/update_config.sh
+++ b/docker/e2e/update_config.sh
@@ -9,20 +9,20 @@ update_toml_key() {
   temp_file="$(mktemp)"
 
   if [[ "$section" = '' ]]; then
-    split -p "^\[" "$file" "${temp_file}_"
+    csplit -s -f "${temp_file}_" "${file}" "/^\[/"
 
     (
-      sed "s/^${key}.*$/${key} = ${value}/" "${temp_file}_aa"
-      rm "${temp_file}_aa"
+      sed "s/^${key}.*$/${key} = ${value}/" "${temp_file}_00"
+      rm "${temp_file}_00"
       cat "${temp_file}_"*
     ) > "$file"
   elif grep -E "^\\[${section}]\$" "$file" > /dev/null; then
     cp "$file" "$temp_file"
-    split -p "\[${section}\]" "$temp_file" "${temp_file}_"
+    csplit -s -f "${temp_file}_" "${file}" "/\[${section}\]/"
 
     (
-      cat "${temp_file}_aa"
-      sed "s/^${key} .*$/${key} = ${value}/" "${temp_file}_ab"
+      cat "${temp_file}_00"
+      sed "s/^${key} .*$/${key} = ${value}/" "${temp_file}_01"
     ) > "$file"
   else
     (

--- a/docker/ledger/Dockerfile
+++ b/docker/ledger/Dockerfile
@@ -3,7 +3,9 @@ FROM many_local/repo_build:latest as builder
 # ==== STAGE 2 ====
 FROM debian:bullseye as runtime
 
-RUN useradd -ms /bin/bash many
+ARG USER_ID
+
+RUN useradd -u ${USER_ID} -ms /bin/bash many
 
 # Contains the ledger.db store.
 VOLUME /persistent

